shared/test/test_animations.rb

```diff
@@ -284,29 +329,25 @@ def test_restore_previous_animation_with_invalid_version
     refute_nil restored_metadata['failed-restore-at'] ? restored_metadata['failed-restore-at'] : restored_metadata['failed_restore_at']
     assert_equal 'bad_version_id', restored_metadata['failed-restore-from-version'] ? restored_metadata['failed-restore-from-version'] : restored_metadata['failed_restore_from_version']
 
-    @api.delete_object(anim)
+    soft_delete(filename)
   end
 
   def test_restore_deleted_animation_with_invalid_version_does_nothing
     filename = @api.randomize_filename('test.png')
     delete_all_animation_versions(filename)
 
     # Create an animation file
-    v1_file_data = 'stub-v1-body'
-    @api.post_file(filename, v1_file_data, 'image/png')
-    assert successful?
+    upload(filename, 'stub-v1-body')
 
     # Delete it.
-    @api.delete_object(filename)
-    assert successful?
+    soft_delete(filename)
 
     # List object versions
     versions_old = @api.list_object_versions(filename)
     assert successful?
 
     # Attempt restore with bad version ID
-    animation_bucket = AnimationBucket.new
-    response = animation_bucket.restore_previous_version(@channel_id, filename, "bad_version_id", nil)
+    response = restore_version(filename, 'bad_version_id')
 
     # Response indicates mot modified
     assert_equal 'NOT_MODIFIED', response[:status]
```

Comment:
I'm considering changing this behavior now that we've written the `GET` fallback - it seems inconsistent.

- When `GET`ting an animation with a bad version, and the latest version is deleted, we return the latest nondeleted version.

- When restoring an animation to a bad version, and the latest version is deleted, we return `NOT_MODIFIED` which leaves the project 'broken' (latest is deleted in S3) but subsequent `GET` operations will fall back on the latest nondeleted version thanks to the new behavior.

So, on bad restore, should we restore the latest nondeleted version to mirror the `GET` behavior?

Reply:
This is hard to answer because it's not clear exactly how we're getting into a situation where the version we are looking for is missing. The question is, in the absence of the requested version, whether to use the latest version or the latest non-deleted version.

One scenario we know of where the requested version id will be missing is the remix scenario, since remixing only copies the latest version, and gives it a new version id. In this case, using the most recent version would suffice, although either could be fine.

In the scenario where a student deletes an animation from the project, but the code still references it, returning latest non-deleted version could be quite confusing. The animation would still appear in their project when run, but it wouldn't be visible in the animations tab. So it seems best to return latest version (rather than latest non-deleted version) in the `GET` case.

In the restore scenario, I can't think of a way it would be bad to restore latest non-deleted. so I don't feel strongly about which option you go with there.

Are there any other scenarios you are aware of to consider?

Based on the scenarios I am aware of, it seems simplest and best to return latest (rather than latest non-deleted) in both cases, which is likely to solve the remix scenario and the majority of problem cases. By contrast, returning latest non-deleted doesn't give a clear advantage, and could lead to confusing behavior described above, at least in the `GET` case.

My $0.02 :-)

Reply:
In the scenario where a student deletes an animation from the project but the code still references it, no API call occurs. The manifest stands between the student code and all API calls - it provides the mapping from the animation name the student code uses to the actual animation filename.

Here's the restore scenario I'm considering:

1. Student remixes a project with animation list `[animA: version source-1]`.
   Remix is still broken, so their new project's manifest reads `[animA: version source-1]` even though it should read `[animA: version remix-1']`.
   History of `animA` on S3 says `[version remix-1]`.

2. Student edits their project and deletes `animA`.
   The new version of their project has manifest `[]`.
   History of `animA` on S3 says `[<delete marker>, version remix-1]`.

3. Student restores their project to its initial version.  The restore source manifest still has the bad version id, and reads `[animA: version source-1]`.

I think the _expected_ behavior, from the student's point of view, is that we restore the latest non-deleted version, and they see `version remix-1` again.  The final state would be:
- Manifest reads `[animA: version remix-3]`.
- History of `animA` on S3 says `[version remix-3, <delete marker>, version remix-1]`
This gives correct results when the client asks for animations by version (future work), or just gets the latest version (current behavior).

Actual behavior right now is that we no-op when the latest version is deleted, giving us this final state.
- Manifest reads `[animA: version source-1]`
- History of `animA` on S3 says `[<delete marker>, version remix-1]`
This leaves the underlying data in a bad state.  Under current client behavior (get latest) we serve a 404 and the user sees the "lost animation" dialog.  Under future client behavior (get version) we serve `version-remix-1` so the result is correct, but we go through the fallback path (and extra S3 calls) on every GET for that animation in the future.

That's the only "restore to bad version id where latest version is deleted" scenario I could come up with - and in that case, I think latest non-deleted is the best option.

Reply:
ahhh, well-reasoned. yes, restoring the latest non-deleted version sounds best! I'm not sure if it matters for the `GET` case, but being consistent and restoring latest non-deleted sounds fine, now that you've explained why my problem scenario isn't a problem.